### PR TITLE
{bio}[CrayGNU/2016.03] GROMACS v5.1.2

### PR DESCRIPTION
**Pull request by gppezzi**
(created using `eb --new-pr`)

requires https://github.com/hpcugent/easybuild-easyblocks/pull/951, ~~https://github.com/hpcugent/easybuild-easyconfigs/pull/2615~~

### CODE REVIEW

**Comment by gppezzi** [comment]
Test report by @gppezzi
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
daint104 - Linux SLES 11_SP3, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.6
See https://gist.github.com/2f2ca455bfd4ba69d003e21c878d13f5 for a full test report.

**Comment by boegel** [comment]
@gppezzi missing deps?

```
EasyBuildError: 'Irresolvable dependencies encountered:
CMake/3.5.0,
Boost/1.60.0-CrayGNU-2016.03-Python-2.7.11,
libxml2/2.9.3-CrayGNU-2016.03'
```

**Comment by boegel** [comment]
@gppezzi also, please mention the required easyblock PR in the description

**Comment by gppezzi** [comment]
@boegel  I think we need to finish the PR [Python-2.7.11-CrayGNU-2015.11](https://github.com/hpcugent/easybuild-easyconfigs/pull/2608) before tackling this Gromacs. Once it is merged, I'll prepare the 2016.03 version based on that.

This Gromacs build also depends on [Boost](https://github.com/hpcugent/easybuild-easyconfigs/pull/2615).

**Comment by boegel** [comment]
:+1:

File: easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    'http://xmlsoft.org/sources/old/'
+]
+sources = [SOURCELOWER_TAR_GZ]
+
+configopts = 'CC="$CC" CXX="$CXX" --with-pic --without-python'
+
+dependencies = [('zlib', '1.2.8')]
+
+sanity_check_paths = {
+    'files': [('lib/libxml2.a', 'lib64/libxml2.a'), ('lib/libxml2.so', 'lib64/libxml2.so')],
```
Comment (boegel) [inline easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb]: please use `'libxml2.%s' % SHLIB_EXT` rather than hardcoding `.so` (twice)

File: easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+toolchain = {'name': 'CrayGNU', 'version': '2016.03'}
+toolchainopts = {'pic': True}
+
+source_urls = [
+    'http://xmlsoft.org/sources/',
+    'http://xmlsoft.org/sources/old/'
+]
+sources = [SOURCELOWER_TAR_GZ]
+
+configopts = 'CC="$CC" CXX="$CXX" --with-pic --without-python'
```
Comment (boegel) [inline easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb]: please add `--with-zlib=$EBROOTZLIB`

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+source_urls = ['ftp://ftp.gromacs.org/pub/gromacs/']
+sources = [SOURCELOWER_TAR_GZ]
+
+configopts = ' -DCMAKE_C_FLAGS="-g -fopenmp -fPIC" -DCMAKE_CXX_FLAGS="-g -fopenmp -fPIC" '
+configopts += ' -DCUDA_NVCC_FLAGS="-arch sm_35" '
+
+installopts = " -j16"
+ 
+skipsteps = ['test']
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: use `runtest = False`? and add a comment above to explain why you skip testing

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+toolchain = {'name': 'CrayGNU', 'version': '2016.03'}
+toolchainopts = {'usempi': True, 'openmp': True}
+
+source_urls = ['ftp://ftp.gromacs.org/pub/gromacs/']
+sources = [SOURCELOWER_TAR_GZ]
+
+configopts = ' -DCMAKE_C_FLAGS="-g -fopenmp -fPIC" -DCMAKE_CXX_FLAGS="-g -fopenmp -fPIC" '
+configopts += ' -DCUDA_NVCC_FLAGS="-arch sm_35" '
+
+installopts = " -j16"
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: this doesn't make much sense? first of all it's hardcoded, second it's not used at all?

if you want to speed up the GROMACS installation, you should make the easyblock aware of `self.cfg['parallel']` (which gets a good value by default, and which you can control with `--parallel`, `parallel = 16` and `maxparallel = 4` (the latter to limit to max parallelism rather than fixing it hard)

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+# License::   MIT/GPL
+# $Id$
+#
+# This work implements a part of the HPCBIOS project and is a component of the policy:
+# http://hpcbios.readthedocs.org/en/latest/HPCBIOS_2012-93.html
+
+name = 'GROMACS'
+version = "5.1.2"
+cudaversion = '7.0'
+versionsuffix = '-cuda-%s' % cudaversion
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: since you're enabling both MPI & OpenMPI, this is still a hybrid build, so we should keep the `hybrid` in the versionsuffix?

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#             George Tsouloupas <g.tsouloupas@cyi.ac.cy>, Kenneth Hoste
+# License::   MIT/GPL
+# $Id$
+#
+# This work implements a part of the HPCBIOS project and is a component of the policy:
+# http://hpcbios.readthedocs.org/en/latest/HPCBIOS_2012-93.html
+
+name = 'GROMACS'
+version = "5.1.2"
+cudaversion = '7.0'
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: make this  `7.0.28`, and use `cudaversion` (or `cudaver`) below to define the dep

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+configopts = ' -DCMAKE_C_FLAGS="-g -fopenmp -fPIC" -DCMAKE_CXX_FLAGS="-g -fopenmp -fPIC" '
+configopts += ' -DCUDA_NVCC_FLAGS="-arch sm_35" '
+
+installopts = " -j16"
+ 
+skipsteps = ['test']
+
+builddependencies = [
+    ('CMake', '3.5.0'),
+    ('cudatoolkit/7.0.28-1.0502.10742.5.1', EXTERNAL_MODULE),
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: use `cudaver` here?

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    ('fftw/3.3.4.3', EXTERNAL_MODULE),
+    ('Boost', '1.60.0', '-Python-2.7.11'),
+    ('libxml2', '2.9.3'),
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: are these all build-only deps?

File: easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb (hunk 2 above)
Comment (lucamar) [inline easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb]: Done

File: easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb (hunk 1 above)
Comment (lucamar) [inline easybuild/easyconfigs/l/libxml2/libxml2-2.9.3-CrayGNU-2016.03.eb]: Done, thanks

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 8 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: Yes, I don't need to load these modules when I run Gromacs

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 5 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: Is "hybrid" really needed in the versionsuffix? Many scientific applications have MPI & OpenMP already and cuda in the versionsuffix already defines a new type of "hybrid" build.

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 8 above)
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: So, these are all linked statically then? If so: OK, just wondering. :)

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 5 above)
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: I guess it's not strictly needed, but that's how we've been tagged the GROMACS build until now.

I'm willing to drop the `hybrid` in the versionsuffix, but then we'll need to be consistent...

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 6 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: Sure, thanks.

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 4 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: We can drop +installopts = " -j16".

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 7 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: Done, thanks.

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 8 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: Not really, CUDA library is linked dynamically of course and also FFTW. However GROMACS builds the following library file "easybuild/software/GROMACS/5.1.2-CrayGNU-2016.03-cuda-7.0/lib64/libgromacs_mpi.so" that defines the library paths needed at runtime, so you just need the latter in your LD_LIBRARY_PATH to run.

Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: OK, so either the CUDA/FFTW libraries are statically linked into `libgromacs_mpi.so`, or it's using RPATH.

Either way: fine by me. :)

Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: I think it uses RPATH. in fact when you build the multicore executable (without Cuda), sometimes you might also need to add -DCMAKE_SKIP_RPATH=YES on the Cray systems: https://redmine.gromacs.org/issues/911

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 5 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: Let's leave `hybrid` in the versionsuffix for the time being and discuss that on a long term timeline.

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb (hunk 3 above)
Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: We used to skip tests as they call the MPI executable gmx_mpi and they would fail systematically on the login nodes, where we usually build: we would need to submit a SLURM job to run the MPI executable on the compute nodes.

Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: @lucamar OK, please include a comment clarifying that in the easyconfig file itself.

Comment (lucamar) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.eb]: @boegel Done, thanks

**Comment by gppezzi** [comment]
Test report by @gppezzi
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
daint104 - Linux SLES 11_SP3, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.6
See https://gist.github.com/bb01b800b41464c5bef436b67b9dadb9 for a full test report.

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.28-hybrid.eb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+description = """GROMACS is a versatile package to perform molecular dynamics,
+ i.e. simulate the Newtonian equations of motion for systems with hundreds to millions of particles."""
+
+toolchain = {'name': 'CrayGNU', 'version': '2016.03'}
+toolchainopts = {'usempi': True, 'openmp': True}
+
+source_urls = ['ftp://ftp.gromacs.org/pub/gromacs/']
+sources = [SOURCELOWER_TAR_GZ]
+
+configopts = ' -DCMAKE_C_FLAGS="-g -fopenmp -fPIC" -DCMAKE_CXX_FLAGS="-g -fopenmp -fPIC" '
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.28-hybrid.eb]: @gppezzi why is this needed? the `toolchainopts` should be changed instead to enable `-fPIC`? and I wouldn't enable `debug` (`-g`) by default?

File: easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.28-hybrid.eb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+configopts += ' -DCUDA_NVCC_FLAGS="-arch sm_35" '
+
+builddependencies = [
+    ('CMake', '3.5.0'),
+    ('cudatoolkit/%s-1.0502.10742.5.1' %cudaversion, EXTERNAL_MODULE),
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-5.1.2-CrayGNU-2016.03-cuda-7.0.28-hybrid.eb]: style nitpicking: space after `%`

**Comment by gppezzi** [comment]
Test report by @gppezzi
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
daint105 - Linux SLES 11_SP3, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.6
See https://gist.github.com/9245089a2c865ed4263e68842a85d671 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 4 out of 4 (4 easyconfigs in this PR)
daint101 - Linux SLES 11_SP3, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.9
See https://gist.github.com/1e87671b7b620c13f081da437eb75475 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 4 out of 4 (4 easyconfigs in this PR)
daint101 - Linux SLES 11_SP3, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.9
See https://gist.github.com/14ba31975a970025a144b93e8125c5e5 for a full test report.

**Comment by gppezzi** [comment]
Test report by @gppezzi
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
daint102 - Linux SLES 11_SP3, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.6
See https://gist.github.com/bfa5121e5ee6df4de5dc1e89e30edfb5 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 4 out of 4 (4 easyconfigs in this PR)
daint104 - Linux SLES 11_SP3, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.9
See https://gist.github.com/301c8cb673e0f9d62e1b19c04e99aaa0 for a full test report.

**Comment by gppezzi** [comment]
@boegel My last test was using https://github.com/hpcugent/easybuild-easyblocks/pull/951

Any other remarks?

**Comment by boegel** [comment]
lgtm

**Comment by boegel** [comment]
Going in, thanks @gppezzi!
